### ISSUE TITLE
Yarn is not deduping certain packages with * dependencies

### ISSUE THREAD

**Issue by nickpape-msft** [yarnpkg/yarn#6695]
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**

Yarn doesn't dedupe @types packages or packages with `*` range dependencies.

**If the current behavior is a bug, please provide the steps to reproduce.**

https://github.com/nickpape-msft/yarn-react-types

Go in there and run `yarn install`. You would expect there to be only the following versions:

```json
"@types/react": "16.4.2",
"@types/react-dom": "16.0.5"
```

You would expect this because the `@types/react` that is installed at the root should satisfy the dependency listed in `@types/react-dom`:

```json
{
    "name": "@types/react-dom",
    "version": "16.0.5",
    "dependencies": {
        "@types/react": "*",
        "@types/node": "*"
    }
}
```

However, you instead get a side-by-side `@types/react` dependency:

```
λ yarn why @types/react
yarn why v1.12.3
[1/4] Why do we have the module "@types/react"...?
[2/4] Initialising dependency graph...
[3/4] Finding dependency...
[4/4] Calculating file sizes...
=> Found "@types/react@16.4.2"
info Has been hoisted to "@types/react"
info This module exists because it's specified in "dependencies".
=> Found "@types/react-dom#@types/react@16.7.6"
info This module exists because "@types#react-dom" depends on it.
Done in 0.09s.
```

```
λ yarn why @types/react-dom
yarn why v1.12.3
[1/4] Why do we have the module "@types/react-dom"...?
[2/4] Initialising dependency graph...
[3/4] Finding dependency...
[4/4] Calculating file sizes...
=> Found "@types/react-dom@16.0.5"
info Has been hoisted to "@types/react-dom"
info This module exists because it's specified in "dependencies".
Done in 0.09s.
```

This would cause you to get TypeScript build errors (the newest types are perhaps using new compiler features, or you end up with duplicate definitions for stuff that is exported from `@types/react`).

**What is the expected behavior?**

Only the following two packages are installed:

```json
"@types/react": "16.4.2",
"@types/react-dom": "16.0.5"
```

**Please mention your node.js, yarn and operating system version.**

Node: 8.9.4
Yarn: 1.12.3
Windows 10

**Comment by nickpape-msft** [comment]
This is causing an issue for some people using our Yeoman generator. As a workaround, we are using resolutions, but since this behavior seems different from other major package managers (and seems to be fundamentally at odds with the @types system, as well as the concept of deduplication), it seemed like a Yarn bug.

https://github.com/SharePoint/sp-dev-docs/issues/2934

**Comment by arcanis** [comment]
Cf this thread: https://github.com/yarnpkg/yarn/issues/3951#issuecomment-316424639

> Yes and no. This is a bit tricky.
> 
> - **Yes**, using `*` means that the hoister is indeed *allowed* to consider that the dependency can be safely merged with any other version if it wants to
> 
> - **However**, it doesn't mean that the hoister is *required* to do so. Semantically speaking, a `dependencies` entry literally means "this package MUST be able to require a package of version $VERSION". There's absolutely no mention about this version being compatible with any other. If you need this behavior, then you have to use `peerDependencies`, which provide the extra guarantee that the selected version MUST be the same as the one provided by the parent package (which is exactly what is asked for in the first post).
> 
> So, to sum up: the hoister is about trimming the fat and avoiding filling your disk with garbage - not ensure your dependencies are correctly setup (that's the responsibility of package authors). Ideally, your code should work with no hoisting, or with super-weird hoisting that would create ten levels of indirection. If you don't, it will eventually break, whether it's in Yarn or in another package manager.
> 
> Of course, we *could* potentially fix this (most likely not without spending some serious time, and breaking other behaviors). But I fear that by doing so, we would actually enable this kind of "We can do whatever we want, the package managers will have to make it work anyway" behavior, which I think is detrimental to everyone in the long term (Yarn, but also other package managers who would have to have the exact same behavior, and package consumers that would have to use sub-par package managers that can't apply aggressive optimizations because some packages do not play by the rules - we've already been hit by that in a few occasions). 😉 
> 
> --- 
> 
> Now, from an implementation viewpoint, here is what causes the behavior described in the OP. First, you have to understand that what we called the "hoister" is actually splitted in two different parts:
> 
> - The **optimizer** currently runs during resolution, and tries to avoid resolving ranges that have already been satisfied by previous resolution requests, but the requests are mostly independants (so for example, in the OP situation, when we resolve `angular`, we have no idea that a sub-dependency has also a dependency on `angular`, and conversely). The result of the optimizer is what is stored inside the lockfile.
> 
> - The **hoister** runs once the resolution (and optimization) has completed, and I think simply tries to merge together packages that share a exact same version, _without any consideration for their original range_.
> 
> So during the first step, both angular versions are resolved to their respective versions, independently from each other, then the hoister tries to merge them, see that their versions don't match, and just doesn't merge them. And as described in the first part of this comment, this is to be expected. If you *want* them to be merged, you would instead use a `peerDependencies` entry, which would not be resolved and wouldn't suffer the issue described here.

**Comment by arcanis** [comment]
tldr: open a PR on DefinitelyTyped to change the dependencies into peer dependencies 🙂

**Comment by pgonzal** [comment]
> tldr: open a PR on DefinitelyTyped to change the dependencies into peer dependencies 

Also, everybody should have a unicorn!  

If you can't obtain a unicorn for some reason (e.g. you need one of the 5000+ NPM packages that are not part of this awesome fantasy), then you should use a different package manager than Yarn.

Ermmm...  wait a sec--  we want people to use Yarn, right?!  🙂

**Comment by milesj** [comment]
Filed a similar issue here: https://github.com/yarnpkg/yarn/issues/6415

**Comment by arcanis** [comment]
> Also, everybody should have a unicorn!

I want to think that opening a PR to remove an undefined behavior from a package should be less work than finding a unicorn. But someone has to do it at some point, I guess 🙂

**Comment by pgonzal** [comment]
But (as far as I know) the current definitions work correctly today with NPM and PNPM, which have two radically different installation strategies. So, in what sense is the behavior "undefined"?

Yarn's interpretation seems to be that if I ask for "~1.2.3", I should always get the absolute latest 1.2.99, even if all the other libraries in my locus asked for "1.2.4". That is wasteful, because we have to install, load, and eval two entire copies of the library. It seems pretty reasonable for the package manager to say "1.2.4 is compatible with what you asked for, so it's more economical to give you that."

@zkochan

**Comment by arcanis** [comment]
> So, in what sense is the behavior "undefined" ? 

Two reasons:

- You have to understand that perfect non-wasteful resolution across a tree is Np-complete, and would require to preemptively download the metadata of all the dependencies of all the releases of all the packages in the tree. We (and that includes pnpm and npm) simply cannot guarantee that all the projects will be "perfectly optimized" across the tree, so while it.

- The definition of "perfect" and "non-wasteful" is entirely objective and whatever implementation we have, someone will complain. You're arguing that `*` means "anything", and that resolving to old release to optimize for space is "better". I guarantee you that someone else will open an issue if this behavior changes and tell us how "dangerous" it is that we don't resolve to latest every time we can because bugfixes are not automatically pulled etc etc. And I'll have to make the same case again.

When I say undefined behavior, I mean it in the C/C++ sense: it's undefined in the sense that nothing defines what happens when two packages from the dependency tree overlap[1]. They might be resolved to the same version, but maybe not. And the behavior can change if we need to.

[1] Except in Plug'n'Play, where it is standardized that regardless of the position of the dependent package in the dependency tree, the exact same range absolutely must resolve to the exact same version (we can afford to do it there because we're not restricted by the hoisting anymore). Still not guarantees on overlapping-but-different ranges, though (because of the two points I detailed above).

> Is that the work needed to make the @types packages compatible with Yarn?

You can work incrementally and fix such packages as they cause issues, but generally yes. You also can send a PR to Yarn to fix the bug that duplicates the resolution, and keep the DefinitelyTyped packages as they are (broken). They'll cause other issues down the road because the core problem won't have been addressed, of course. 

If you want a hotfix, the `resolution` field is the best answer. Just tell Yarn that all versions of your @types package must absolutely resolve to a version of your choice.

**Comment by milesj** [comment]
@arcanis Correct me if I'm wrong, but doesn't `*` mean any? So why would you need to preemptively download all the past versions as well? Wouldn't `package@latest` be enough? Or simply, whatever the tree has resolved to be the "latest"?

**Comment by arcanis** [comment]
> So why would you need to preemptively download all the past versions as well?

Because a "perfect" tree (given any subjective metric) doesn't necessarily use all the latest versions.

For example a "perfect" tree (I hate this term so much) where the project depends on A@1.0.0 and B@* would maybe choose the older release if given the choice between B@1.0.0 (which itself depends on A@1.0.0) and B@2.0.0 (which itself depends on B@2.0.0), because it would only require to install two packages (vs three with the more recent package).

**Comment by mikl** [comment]
The most confusing part of this problem is that it behaves non-deterministically. I had been using Yarn for a TypeScript project for months, and I had been lucky to not hit this problem, until one random afternoon it blows up in your face and starts installing mutually incompatible versions all over the place.

If Yarn is not able to handle`"*"`  versions the way npm does, it should at least warn the user that this syntax is not supported they way the npm ecosystem expects.

The way things stand at the moment, Yarn cannot safely be used in combination with TypeScript (since almost any TS project is bound to use `@types/` packages, and the vast majority of them have `"*"` versions in their dependencies). Sooner or later, this is going to lead to strange compiler errors.

**Comment by arcanis** [comment]
> If Yarn is not able to handle"*" versions the way npm does, it should at least warn the user that this syntax is not supported they way the npm ecosystem expects.

I think you're missing the point, which is that this syntax is supported the exact same way as npm in that we make the same guarantees. Or rather that *neither of us make the guarantees that @types is relying on.* If the case you describe works with npm it's purely thanks to an implementation detail, and while it would make sense to do something similar for the sake of optimizing the tree a bit more *relying on a package manager doing so is invalid and broken*. Which kinda downgrade the priority of fixing it on my side (anyone is welcome to open a PR, though).

As much as I like TS and the DefinitelyTyped project, it's quite clear that they didn't properly consider the semantic implications behind using regular dependencies. Trust me it's as annoying for me to explain the same problem multiple times (without anyone starting a discussion to fix that on the TS side) as it is for you to experience it ...

**Comment by mikl** [comment]
It may be that npm has not explicitly committed themselves to supporting this usage, but as long as it keeps working, I think the odds are fairly slim that a community effort like DefinitelyTyped will bother to rewrite the dependencies of thousands of packages, just to make things work with Yarn.

I understand it's annoying, but considering the positively herculean effort in trying to get the DT community to change, to support a minority package manager like Yarn, I don't think it's particularly surprising that nobody has volunteered yet.

Don't get me wrong, I really like Yarn and I admire all the people who've invested their time in pushing the npm ecosystem forwards by creating and maintaining an alternative to npm. But not so much that I want to spend weeks, if not months, trying to get all of DefinitelyTyped to change.

In fact, given the growing popularity of TS, I think if npm will find themselves having to support this usage. A new version of npm that doesn't work with TypeScript would go over like a lead balloon.

**Comment by octogonz** [comment]
> I think the odds are fairly slim that a community effort like DefinitelyTyped will bother to rewrite the dependencies of thousands of packages, just to make things work with Yarn.

I talked to the compiler owners about this a couple years ago, and they eventually convinced me that the `*` versions were the best possible design, even though it seems strange.

The basic issues are:
1. The `@types` version numbers are meant to have the same major/minor as the underlying library, since otherwise it would be extremely confusing to figure out which one you need
2. This means a patch-level increment can have breaking API changes (e.g. if they typings strategy was improved). 
3. Thus `@types` packages do NOT follow SemVer, and it's a huge mistake to use ranges such as `^1.2.3`. The only sensible forms are locked `1.2.3` or star `*` (star meaning it's someone else's responsibility to choose the version)
4. If every `@types` package uses locked versions for its dependencies, the entire graph would become locked, meaning someone would need to centrally coordinate huge sets of typings that are meant to go together. This is infeasible.
5. Also, locked dependencies don't make sense for ambient typings (e.g. node or dom apis), which describe a global environment for the application, and are merely the context in which a library runs.
6. Alternatively, if the `@types` packages use star dependencies, then it creates a multivariable equation that someone needs to solve, i.e. tinker with the versions to find a set of typings that compile without syntax errors and match the underlying js libraries. The end application project is in the best position to solve this equation.

Part of the problem is that the typings are maintained by volunteers who are typically not part of the release process for the underlying library, and often unable to deliver a quality bar beyond whatever they need to get their immediate problem unblocked. This is why `@types` packages encounter a lot of breaks that don't seem to come up when library maintainers ship typings as part of the official package.

Despite all these limitations, DefinitelyTyped works amazingly well in practice, at least from the perspective of end users who consume `@types` packages without being exposed to the sometimes tricky process of how they are produced.

And I agree that NPM/PNPM work very reliably with star dependencies. Theory aside, Yarn clearly has frequent problems that the other package managers do not seem to encounter in practice.

**Comment by arcanis** [comment]
> I talked to the compiler owners about this a couple years ago, and they eventually convinced me that the `*` versions were the best possible design, even though it seems strange.

I don't have anything against `*` - what I don't totally understand is the reason why those dependencies aren't *peer*Dependencies rather than regular dependencies. If it is the job of someone else to pick the version they should resolve to, then it seems to me that they clearly match the use case for peer dependencies - since they precisely allow the parent to influence the children.

If `@types/react-dom` was listing `@types/react` as a peer dependency, then it would 1/ resolve this problem (because `@types/react` would by definition have the same resolution as the one provided by the package depending on `@types/react-dom`), and 2/ make it explicit to the consumer of `@types/react-dom` that they also need to depend on `@types/react`.

> I understand it's annoying, but considering the positively herculean effort in trying to get the DT community to change, to support a minority package manager like Yarn, I don't think it's particularly surprising that nobody has volunteered yet.

Yep I totally understand - I spend myself a lot of time advocating for good practices and submitting [PRs](https://user-images.githubusercontent.com/1037931/53919899-48460880-4063-11e9-9026-f1f131858894.png) to fix invalid dependency listings as I stumble upon them. While it's not always the most interesting PRs, I feel like it's important to progressively put back some strictness in the ecosystem. Javascript popularity grew fast, and I think we're reaching a pivotal moment where we need to strengthen our foundations (and tools) rather than doubling down whatever the cost.

---

Anyway, regardless of my personal opinion, I would be willing to review a PR that would help improve this use case and add a regression test to protect against. Any brave soul out there? 🙂

**Comment by octogonz** [comment]
> I don't have anything against `*` - what I don't totally understand is the reason why those dependencies aren't _peer_Dependencies rather than regular dependencies. If it is the job of someone else to pick the version they should resolve to, then it seems to me that they clearly match the use case for peer dependencies - since they precisely allow the parent to influence the children.

In the past (e.g. at least as recently as 4.x), NPM's behavior when installing peer dependencies was so misguided that peer dependencies were effectively useless.  (Basically, if you had a dependency chain App->MiddleWare->Library, the MiddleWare package was unable to specify versions for peer dependencies declared by the Library.)  During this era, people got frustrated with peer dependencies, and they were widely deprecated.  PNPM and Yarn don't have this flaw, and it may be the case that NPM eventually fixed it as well.  Recently people seem to be using peer dependencies more, and I haven't heard as many complaints about them.

@arcanis if all of the `@types` packages declared **peer** dependencies using `*` for the version, would Yarn do the right thing?  

@zkochan would PNPM handle this okay?

> I understand it's annoying, but considering the positively herculean effort in trying to get the DT community to change, to support a minority package manager like Yarn, I don't think it's particularly surprising that nobody has volunteered yet.

I wouldn't personally consider this to require a "herculean effort."  If someone could persuasively demonstrate that a different design would significantly improve the DT experience, the code it is pretty much all in one big master branch.  It's probably not a big deal to make a sweeping change.

**Comment by arcanis** [comment]
> if all of the @types packages declared peer dependencies using `*` for the version, would Yarn do the right thing?

I think so - the main thing being that it would be the responsibility of the application authors to add those packages to their dependencies (whereas now if you add `@types/react-dom` you get `@types/react` "for free" in a completely unintended way). So probably major version bump (or patch since types packages don't follow semver).

Btw, I think I'd be open to make @types packages use exact versions by default when running `yarn add`, if that can help you (since you mention they never respect semver).

**Comment by zkochan** [comment]
> @zkochan would PNPM handle this okay?

I believe the solution described by @arcanis will work with pnpm as well. Pnpm supports peer dependencies and resolves them correctly.

> Btw, I think I'd be open to make @types packages use exact versions by default when running yarn add (instead of the regular ^), if that can help you (since you mention they never respect semver)

off topic, I'd do that for everything :-)

**Comment by octogonz** [comment]
> > Btw, I think I'd be open to make @types packages use exact versions by default when running yarn add (instead of the regular ^), if that can help you (since you mention they never respect semver)
> 
> off topic, I'd do that for everything :-)

`rush add` by default writes `~1.2.3` instead of `^1.2.3`, and for `@types` packages it uses `1.2.3`.  We've been doing this for a while, and people seem pretty happy with this behavior.

> > if all of the `@types` packages declared peer dependencies using `*` for the version, would Yarn do the right thing?
> 
> I think so - the main thing being that it would be the responsibility of the application authors to add those packages to their dependencies

Proposal:  Maybe Yarn/PNPM could help out by adding a new command line option `--install-types-as-peers`. With this option enabled, when the package manager encounters this:

```
{
  "name": "@types/lib1",
  "version": "1.2.3",
  "dependencies": {
    "@types/lib2": "*"
  }
}
```

...it would get installed as if it had been written like this:

```json
{
  "name": "@types/lib1",
  "version": "1.2.3",
  "peerDependencies": {
    "@types/lib2": "*"
  }
}
```

And when the package manager encounters this:

```json
{
  "name": "lib3",
  "version": "1.2.3",
  "dependencies": {
    "@types/lib1": "1.2.3"
  }
}
```

...it would get installed as if it had been written like this:

```json
{
  "name": "lib3",
  "version": "1.2.3",
  "dependencies": {
    "@types/lib1": "1.2.3"
  },
  "peerDependencies": {
    "@types/lib2": "*"
  }
}
```


This would allow people to opt-in to this new model and test the waters. If many professional code bases find that this approach works reasonably well, then we could use that evidence to persuade DefinitelyTyped to convert all their packages to adopt it.

**Comment by milesj** [comment]
Unless I missed something in this comment chain, everyone keeps talking about the consumer installing all the `@types` dependencies that are required in their application. But this doesn't seem to take into account all the NPM packages that includes `@types` in their `dependencies`, which I would assume is quite a bit, and mostly fixed versions and not `*`.

For example, just look at `@types/react`, which is being depended on more than 2000 times. https://www.npmjs.com/package/@types/react

I don't see how `peerDependencies` is the solution here. The consuming application *should not* have to install `@types` for literally every NPM package in it's tree that requires them.

**Comment by octogonz** [comment]
> But this doesn't seem to take into account all the NPM packages that includes `@types` in their `dependencies`, which I would assume is quite a bit, and mostly fixed/caret versions, and not `*`.

I just did a quick audit of the 2,131 NPM packages consumed by a real production repo.  This includes internal packages and external packages from npmjs.com, and a mixture of browser libraries and NodeJS libraries.

Ignoring `devDependencies`, the `dependencies` sections contain a total of 147 references to `@types` packages.  Of those, 121 are locked versions and 3 are star versions.  Only 24 of them are SemVer ranges, and here's the list:

<table>
<tr><td><b>range</b></td><td><b>dependency</b></td><td><b>consuming package</b></td></tr>
<tr><td>^1.0.7</td><td>@types/applicationinsights-js</td><td>@mskeros/chisel</td></tr>
<tr><td>^15.6.1</td><td>@types/create-react-class</td><td>@ms/project-web-shared</td></tr>
<tr><td>^1.2.3</td><td>@types/d3-shape</td><td>@uifabric/charting</td></tr>
<tr><td>^2.1.0</td><td>@types/d3-time-format</td><td>@uifabric/charting</td></tr>
<tr><td>^2.2.29</td><td>@types/invariant</td><td>react-dnd</td></tr>
<tr><td>~1.4.0</td><td>@types/jju</td><td>@microsoft/node-core-library</td></tr>
<tr><td>^4.14.116</td><td>@types/lodash</td><td>@mskeros/chisel</td></tr>
<tr><td>^4.14.109</td><td>@types/lodash</td><td>react-dnd</td></tr>
<tr><td>^4.14.109</td><td>@types/lodash</td><td>react-dnd-html5-backend</td></tr>
<tr><td>^0.0.32</td><td>@types/mz</td><td>@pnpm/link-bins</td></tr>
<tr><td>^0.0.32</td><td>@types/mz</td><td>@pnpm/package-bins</td></tr>
<tr><td>^9.6.5 || 10</td><td>@types/node</td><td>@pnpm/link-bins</td></tr>
<tr><td>^9.4.0 || 10</td><td>@types/node</td><td>@pnpm/logger</td></tr>
<tr><td>^8.0.47</td><td>@types/node</td><td>adal-node</td></tr>
<tr><td>^10.3.0</td><td>@types/node</td><td>react-dnd</td></tr>
<tr><td>^15.5.3</td><td>@types/prop-types</td><td>react-dnd</td></tr>
<tr><td>^0.25.20</td><td>@types/ramda</td><td>@pnpm/link-bins</td></tr>
<tr><td>^16.3.14</td><td>@types/react</td><td>react-dnd</td></tr>
<tr><td>^16.3.16</td><td>@types/react</td><td>react-dnd-html5-backend</td></tr>
<tr><td>^16.0.5</td><td>@types/react-dom</td><td>react-dnd</td></tr>
<tr><td>^1.0.1</td><td>@types/react-dom-factories</td><td>@ms/project-web-shared</td></tr>
<tr><td>^0.2.2</td><td>@types/shallowequal</td><td>react-dnd</td></tr>
<tr><td>^0.2.2</td><td>@types/shallowequal</td><td>react-dnd-html5-backend</td></tr>
</table>

My guess is that only a few of these packages are actively churning. Most have either (1) not had a SemVer-compatible release in a long time, or else (2) are very well-behaved APIs such as **lodash.**

Other may have different experiences. What I observed was that during the evolution of this repo, we spent a long time dealing with frustrating breaks caused by patch-increments for DefinitelyTyped packages, until we eventually managed to lock down a consistent set of version numbers for our application and all its dependencies.  That's why the set of `^` and `~` ranges is now very small.

To summarize:
- SemVer ranges cause annoying breaks for `@types` packages
- If every library locks uses versions in `dependencies`, then you get massive fragmentation of side-by-side installs, including problems such as the compiler getting confused by two different typings for the same API
- The reliable solution is for each library to lock its own tree of versions using `devDependencies`.  This allows consumers of a library to arrive at their own solution.  
- (You can get away with other approaches for certain well-behaved edge cases such as above, but those approaches break down when propagated across a large enough dependency graph.)
- If you're looking for an "ideal" solution, make a PR to add the .d.ts files to the upstream JavaScript package, and get the owners to commit to supporting TypeScript.  `@types` is merely a hacky step towards the right thing.

**Comment by milesj** [comment]
Ideally, if all of these libraries used `*` versions for `@types` in `dependencies`, would that alleviate most of these install issues? That way, `latest` is always used when installing transitive, and applications can lock their `@types` versions as they will.

**Comment by arcanis** [comment]
Note that the problem isn't about resolving to `latest`. It's just that, at a basic level, Yarn will only ever resolve one range to one version, regardless of the position of the package in the dependency tree (this is different from npm which might resolve the same range to different versions depending on the package location on the dependency tree, for the better or worse). This mechanism is a cornerstone of our architecture, and unlocks various core optimizations.

From an implementation perspective, imagine that there is a `Map` whose key is a range and the value is the resolution. When we see a package depending on something, we check whether the range has been resolved and, if it hasn't, we enqueue it to be resolved later. Since the ranges resolve to the same versions regardless of the package that required them, we don't really factor the tree location when considering which version should be used for a specific range<sup>1</sup> - hence why `^15.0.0` and `*` get different result<sup>2</sup>.

> Could this also be solved by a Yarn v2 plugin?

Technically you could have a resolver that would bind itself on `@types/*` resolutions and would convert the `dependencies` into `peerDependencies` on the flight, but that definitely wouldn't be something I'd recommend. Plugins should extend the spec, not replace it - especially when more correct solutions have been identified.

---

<sup>1</sup> This isn't entirely true in that we do try to apply some optimizations when possible - but generally speaking you should assume that we can't

<sup>2</sup> If you were to suggest that we simply reuse the result from `^15.0.0` when resolving `*`, consider that the resolution order matters: what if `*` is resolved before `*`? Fixing this would be quite hard, and a lot of effort to circumvent a problem that originates in the first place in how packages are defined.

**Comment by gitowiec** [comment]
Hi, I am looking for a practical way to cope with this issue in monorepo. Could You help? 
I added such lines to `.yarnrc`
```
save-prefix ""
--install.check-files true
--add.check-files true
--remove.check-files true
--install.frozen-lockfile true
--add.frozen-lockfile true
--remove.frozen-lockfile true
```
It helped me when the @types/react dependency is hoisted, but when Yarn decides to keep it nested inside specific dependency package's `node_modules` (eg @types/redux-form) I still get multiple TypeScript errors during compilation similar to this:
```
TS2717: Subsequent property declarations must have the same type.  Property 'a' must be of type 'DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>', but here has type 'DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>'.
```
How to resolve this issue in monorepo? I will add that Yarn's "resloutions" don't work in monorepos :(

**Comment by mikl** [comment]
@gitowiec I think the only solution atm. is to switch to npm with Lerna. Maybe Tink in the future.

**Comment by gitowiec** [comment]
@mikl Thanks for the reply. But your answer makes me sad. We have a small devs team and this issue is pushing us back. Until now I "fixed" this problem by manually editing yarn.lock. It helped when packages where hoisted from `packages/ui` workspace to the workspaces root.

**Comment by octogonz** [comment]
We've been using [PNPM 3.x](https://github.com/Microsoft/web-build-tools/pull/1210/files#diff-29e04870d678e6105642e71f416b522a) which introduced a `--resolution-strategy=fewer-dependencies` option that so far seems to solve this problem.

**Comment by rjmunro** [comment]
From my reading of what dependencies, devDependencies and peerDependencies are supposed to be (Ignoring bugs in yarn, npm etc.) `@types/...` packages should:

* Always be in `devDependencies` for normal projects or libraries as they are only needed for compilation, not for runtime.
* Always be in peerDependencies for `@types/` that depend on each other,  because it is never valid to have different versions installed at the same time. Currently the typescript compiler cannot cope with such a setup, and it probably wouldn't make sense if it did.
* Consider peer-depending on the thing whose types they are defining, if installing from npm is the only sane way to get that thing.

**Comment by octogonz** [comment]
>     * Always be in peerDependencies for `@types/` that depend on each other,  because it is never valid to have different versions installed at the same time. 

Well, it is never valid for ambient declarations. For normal declarations it is valid in cases where different versions of the underlying library were installed (and are not being used interchangeably).

**Comment by eps1lon** [comment]
> Always be in peerDependencies for @types/ that depend on each other, because it is never valid to have different versions installed at the same time. Currently the typescript compiler cannot cope with such a setup, and it probably wouldn't make sense if it did.

Declarations published under `@types/` always declare their dependencies as actual `dependencies` not as `peerDependencies`. I would agree that it would be better to have those as peers but as long as the biggest types repo is using direct dependencies we should make it work for those.

**Comment by milesj** [comment]
@rjmunro 

> Always be in devDependencies for normal projects or libraries as they are only needed for compilation, not for runtime.

This just isn't true. If a library uses a type from a `@types` package in their public API (argument type, return type, etc), then the `@types` package must be a normal dependency, otherwise the TS compiler will fail.

> Always be in peerDependencies for @types/ that depend on each other, because it is never valid to have different versions installed at the same time. Currently the typescript compiler cannot cope with such a setup, and it probably wouldn't make sense if it did.

What about `@types` packages that have different types between major versions? Using `*` as a peer could theoretically pull in a much higher version than the app/library in question. But this is definitely something `tsc` should resolve.

**Comment by Andrea-Arguello** [comment]
I ran into this after I tried upgrading a new library, and it would just write it under the old one (i.e., the package was:
```
react-native-animatable@^1.2.4:
  version "1.3.2"
  resolved "https://registry.yarnpkg.com/react-native-animatable/-/react-native-animatable-1.3.2.tgz#4783ee1a73dc98815aef234ce6b819f80bfe7d80"
  dependencies:
    prop-types "^15.5.10"
```
instead of
```
react-native-animatable@^1.3.3:
  version "1.3.3"
  resolved "https://registry.yarnpkg.com/react-native-animatable/-/react-native-animatable-1.3.3.tgz#a13a4af8258e3bb14d0a9d839917e9bb9274ec8a"
  integrity sha512-2ckIxZQAsvWn25Ho+DK3d1mXIgj7tITkrS4pYDvx96WyOttSvzzFeQnM2od0+FUMzILbdHDsDEqZvnz1DYNQ1w==
  dependencies:
    prop-types "^15.7.2"
```
I tried to upgrade to the specific version and wound up with _both_ in the yarn.lock file. [According to the documentation, ](https://classic.yarnpkg.com/en/docs/cli/dedupe/) deduping is already covered (or at least supposed to be covered) by warn install. What that did instead was this:
```
react-native-animatable@^1.2.4, react-native-animatable@^1.3.3:
  version "1.3.3"
  resolved "https://registry.yarnpkg.com/react-native-animatable/-/react-native-animatable-1.3.3.tgz#a13a4af8258e3bb14d0a9d839917e9bb9274ec8a"
  integrity sha512-2ckIxZQAsvWn25Ho+DK3d1mXIgj7tITkrS4pYDvx96WyOttSvzzFeQnM2od0+FUMzILbdHDsDEqZvnz1DYNQ1w==
  dependencies:
    prop-types "^15.7.2"
```
And, to my horror, did the same with a bunch of other libraries.

**Comment by milesj** [comment]
I started getting around this by running this alias every time I messed with deps.
```
alias yeet='npx yarn-deduplicate yarn.lock && yarn install'
```
